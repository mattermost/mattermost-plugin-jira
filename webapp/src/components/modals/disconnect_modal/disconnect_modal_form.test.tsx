// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {shallow} from 'enzyme';

import DisconnectModalForm from './disconnect_modal_form';

describe('components/DisconnectModalForm', () => {
    const baseActions = {
        closeModal: jest.fn().mockResolvedValue({}),
        sendEphemeralPost: jest.fn().mockResolvedValue({}),
        disconnectUser: jest.fn().mockResolvedValue({}),
    };

    const baseProps = {
        ...baseActions,
        visible: true,
        theme: {},
        connectedInstances: [],
    };

    test('should match snapshot', () => {
        const props = {...baseProps};
        const wrapper = shallow<DisconnectModalForm>(
            <DisconnectModalForm {...props}/>,
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('should close modal and send ephemeral post on submit success', async () => {
        const closeModal = jest.fn().mockResolvedValue({});
        const sendEphemeralPost = jest.fn().mockResolvedValue({});
        const disconnectUser = jest.fn().mockResolvedValue({});

        const props = {
            ...baseProps,
            closeModal,
            sendEphemeralPost,
            disconnectUser,
        };
        const wrapper = shallow<DisconnectModalForm>(
            <DisconnectModalForm {...props}/>,
        );

        wrapper.instance().handleInstanceChoice('', 'https://something.atlassian.net');
        expect(wrapper.state().selectedInstance).toEqual('https://something.atlassian.net');

        wrapper.instance().submit({preventDefault: jest.fn()});
        await Promise.resolve();

        expect(disconnectUser).toHaveBeenCalledWith('https://something.atlassian.net');
        expect(sendEphemeralPost).toHaveBeenCalledWith('Successfully disconnected from Jira instance https://something.atlassian.net');
        expect(closeModal).toHaveBeenCalled();
    });

    test('should show error on submit fail', async () => {
        const closeModal = jest.fn().mockResolvedValue({});
        const sendEphemeralPost = jest.fn().mockResolvedValue({});
        const disconnectUser = jest.fn().mockResolvedValue({error: 'Error disconnecting'});

        const props = {
            ...baseProps,
            closeModal,
            sendEphemeralPost,
            disconnectUser,
        };
        const wrapper = shallow<DisconnectModalForm>(
            <DisconnectModalForm {...props}/>,
        );

        wrapper.instance().handleInstanceChoice('', 'https://something.atlassian.net');
        expect(wrapper.state().selectedInstance).toEqual('https://something.atlassian.net');

        wrapper.instance().submit({preventDefault: jest.fn()});
        await Promise.resolve();

        expect(disconnectUser).toHaveBeenCalled();
        expect(sendEphemeralPost).not.toHaveBeenCalled();
        expect(closeModal).not.toHaveBeenCalled();

        expect(wrapper.state().error).toEqual('Error disconnecting');
    });
});
