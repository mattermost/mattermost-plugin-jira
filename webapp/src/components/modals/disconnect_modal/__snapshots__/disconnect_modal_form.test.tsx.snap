// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/DisconnectModalForm should match snapshot 1`] = `
<form
  onSubmit={[Function]}
  role="form"
>
  <ModalBody
    style={
      Object {
        "backgroundColor": undefined,
        "color": undefined,
        "padding": "2em 2em 3em",
      }
    }
  >
    <ReactSelectSetting
      onChange={[Function]}
      options={Array []}
      theme={Object {}}
    />
  </ModalBody>
  <ModalFooter
    style={
      Object {
        "padding": "2rem 15px",
      }
    }
  >
    <FormButton
      btnClass="btn-link"
      defaultMessage="Cancel"
      disabled={false}
      extraClasses=""
      onClick={[Function]}
      savingMessage="Creating"
      type="button"
    />
    <FormButton
      btnClass="btn btn-primary"
      defaultMessage="Disconnect"
      disabled={true}
      extraClasses=""
      saving={false}
      savingMessage="Creating"
      type="submit"
    />
    <p
      className="error-text"
    />
  </ModalFooter>
</form>
`;
